### ISSUE TITLE
A button (or automatic behavior) to clear/flush the the VRAM after swapping

### ISSUE THREAD

**Issue by eniora** [C0untFloyd/roop-unleashed#89]
**Describe the bug**
When running gradio roop for multiple video swaps (actually a few, like 3-4 swaps), the VRAM keeps filling up with every swap even on a 3090 with 24GB of VRAM, tried with 6-8 threads even on a 3090 and this keeps happening (although happens much slower and VRAM isn't filled fast with 6-8 threads on a 3090), this wasn't a big issue with old roop unleashed (e.g. 2.0.3) or the original roop (to be fair it happened but not as aggressive as with gradio roop unleashed). For that reason, is it possible to add a button to clear/flush the VRAM of no longer needed data after swapping (or do that automatically after swapping)? For example some SD A1111 extensions have such a button to clear VRAM of the data they use and it's so helpful. I always find myself completely closing roop and reopening it to start fresh which speeds up the swapping almost twice.

**To Reproduce**
1. Add any source image
2. Add a video (1-2 min or so, 720 or 1080p)
3. Swap it
4. Repeat 2-3 times with another source image and another video, the VRAM keeps filling up and even after it finishes swapping it doesn't go lower at all.

**Details**
What OS are you using?
- [ ] Linux
- [ ] Linux in WSL
- [X] Windows 
- [ ] Mac

Are you try to use a GPU?
- [ ] No. I am not using the `---gpu` flag
- [x] NVIDIA
- [ ] AMD
- [ ] Intel
- [ ] Mac

**Sanity Check**
- [X] I have the latest code from the github repository
- [X] I have followed the installation guide

**Comment by Freddy22119** [comment]
I am also experiencing this.

**Comment by TheBigSmokey** [comment]
Just checked this on my pc as well. I'm experiencing the same thing. Using Nvidia Cuda.

**Comment by eniora** [comment]
@TheBigSmokey Yah I have 12GB VRAM on my other PC and after 1-2 swaps it becomes very slow due to filled up VRAM, like the first swap is ~15-20FPS then after swapping another video and getting back to the first as a test, it's down to 5-10FPS with full VRAM usage, and when VRAM is filled the card isn't fully used as compared to the first swap (I can tell not only because of load % but the temperature is also low compared to the first swap). VRAM can only be cleared by completely restarting roop. I know models has to be loaded into VRAM but we're talking about the temp data that gets added every time a face or video is loaded and I believe that can be cleared by a button or automatically after each swap.

**Comment by Freddy22119** [comment]
Yeah I did a quick test of a 10 second video to see if I was going insane. The first test took ~1 minute, then 2, then 4, then 6, and then 9. I changed no settings, I just kept running the same video to see what was happening.

**Comment by eniora** [comment]
@C0untFloyd I think I discovered something. It's about how many threads you set in settings. If I set to 10 threads on my 3090, VRAM usage doesn't really get full quickly but for the default I always use on my 3090 which is 20-24 threads, each thread roughly fills 1GB of VRAM and the more I swap it keeps filling until it gets to 24GB (if I have it set to 24 threads), now 24 threads is much faster than setting it to 10 to avoid this, so it would be nice if we can clear the VRAM after each run. For 8-12GB vram cards even setting it to 6-8 threads isn't really helpful as it can still fill up rather quickly.

**Comment by C0untFloyd** [comment]
Can you point me to a A1111 extension which does this via Button? I did some research a while ago and it seems this is a problem many people have when using cuda tensors. The old roop had a release_resources function, I will reinstate this and call it even between video processing. Let's see if that helps, I will also try the suggestions from this page:
https://saturncloud.io/blog/how-to-clear-gpu-memory-after-pytorch-model-training-without-restarting-kernel/
It's very likely the enhancers have some memory leaks, they are mostly a mess and need to be cleaned up.
 
On the other hand a button (in settings?) might be not such a bad idea because you could have a run with everything enabled, causing all models etc. to be loaded onto the GPU. Even if you would disable everything, those models would still be in memory as those are kept for faster startup.

**Comment by eniora** [comment]
There was an extension that had a clear VRAM button for the models it uses but I installed it long ago and couldn't find it yesterday when I searched for it. Anyway here's a similar one that has a button to clear the models from VRAM:
https://github.com/continue-revolution/sd-webui-animatediff
Here are the buttons it uses for this:
![Screenshot 2023-08-13 124050](https://github.com/C0untFloyd/roop-unleashed/assets/116663473/dc42f8c9-262d-4cbf-aa90-6fb081079d48)

**Comment by Siraj-HM** [comment]
For some reason the new gradio ui unleashed is way too slow compared to the OG unleashed one. I used to get 20 frame per seconds on my 3070 8GB VRAM. Now it is taking ~5-6s per frame for the same video.

**Comment by eniora** [comment]
@Siraj-HM On one of my PCs which has a 1070 8GB I managed to get almost the same speed as with the old unleashed, just make sure it's set to 6 threads in settings and for now restart roop after every few swaps to maintain the speed. 6 threads is a must for 8GB cards and for 12GB it should be set to 10 threads or so.

**Comment by C0untFloyd** [comment]
> Anyway here's a similar one that has a button to clear the models from VRAM:

It unloads its own Model from GPU to CPU and calls a base cleanup function of A1111.
The cleanup function does almost the same as the original roop back then.
I  added it now and also re-implemented the memory limit function from versions ago. I didn't test this extensively but it looked like the memory is freed after each batch-processing run. It's committed now, feel free to test it yourself.

Unloading the models to CPU  would be an improvement too, however most of them are treated as black boxes inside of libraries which care about unloading/loading themselves. This is something for later I guess.

**Comment by lysxelapsed** [comment]
Seems to change max_threads instead of memory_limit

**Comment by eniora** [comment]
I tested on multiple videos (around 15 vids, some while using the enhancer and some without) and did some image batch process swapping, all on a 3090. Max VRAM usage I saw was 22GB, and after each run it gets back to 20GB, then again ~22GB during processing and back to 20GB etc. Didn't see it getting completely full and no slowdowns at all, before the update it was getting full quickly and getting much slower. All these tests were with 24 threads, setting it to 20 threads didn't make much difference (with 8 threads the max I saw was 12.7GB usage). I barely know anything about this stuff and how model loading to GPU works but it doesn't clear the VRAM fully after each run (doesn't only keep the necessary models loaded without the swapping data which should be around 3GB as I can tell after adding the first source and target after starting roop), I know it's most likely not that simple and you will probably lecture me 😁. Anyway what's important is that overall seems like the issue is fixed, at least for my 3090. Will try to test on my other PC with 8GB VRAM tonight, if not will test tomorrow. Meanwhile I am looking forward to see how the performance is now for the others who commented here about this issue. Thanks a lot for all these updates, you're really doing an awesome job.

And yah like @lysxelapsed said, memory limit is currently tied to GPU threads instead of memory_limit in the config file.
